### PR TITLE
Add option to automatically find category variables (searchable dimensions)

### PR DESCRIPTION
**Pull request by geometrikz**
Changes:

- Add option to automatically find and tag searchable dimensions by

1. Approximately counting the distinct number of rows for dimensions
2. Tagging dimensions with <=100 unique fields as "searchable"

### CODE REVIEW

File: metrics_layer/cli/seeding.py [hunk 1]
```diff
@@ -56,7 +56,6 @@ def __init__(
             "CHAR": "string",
             "CHARACTER": "string",
             "STRING": "string",
-            "TEXT": "string",
```
Comment (geometrikz) [inline metrics_layer/cli/seeding.py:59]: These are duplicate keys in the dictionary so I removed them.

File: metrics_layer/cli/seeding.py [hunk 2]
```diff
@@ -106,7 +105,6 @@ def __init__(
             "double precision": "number",
             "numeric": "number",
             "float": "number",
-            "real": "number",
```
Comment (geometrikz) [inline metrics_layer/cli/seeding.py:109]: These are duplicate keys in the dictionary so I removed them.

File: metrics_layer/cli/seeding.py [hunk 3]
```diff
@@ -304,12 +326,38 @@ def make_fields(self, column_data: str):
...
+                cardinality = cardinality["CARDINALITY"].values[0]
+                if cardinality <= 100:
+                    field["searchable"] = True
             else:
                 field["field_type"] = "dimension"
                 field["type"] = metrics_layer_type
             fields.append(field)
         return fields
 
+    def column_cardinality_query(self, column_name: str, schema_name: str, table_name: str) -> str:
```
Comment (geometrikz) [inline metrics_layer/cli/seeding.py:345]: This function implements the column cardinality query."

AFAIK, Postgres doesn't have an approximate count. We need to decide whether or not to disable for Postgres of keep it in, taking note of the performance/cost of count distinct over all dimensions.

File: metrics_layer/cli/seeding.py [hunk 4]
```diff
@@ -304,12 +326,38 @@ def make_fields(self, column_data: str):
                 field["type"] = "time"
                 field["timeframes"] = ["raw", "date", "week", "month", "quarter", "year"]
                 field["datatype"] = metrics_layer_type
+            elif metrics_layer_type == "string" and auto_tag_searchable_fields:
```
Comment (geometrikz) [inline metrics_layer/cli/seeding.py:329]: Here I check for all string field types. Is there another filter we can apply to target categorical fields better?

File: metrics_layer/cli/seeding.py [hunk 5]
```diff
@@ -304,12 +328,71 @@ def make_fields(self, column_data: str):
...
+            Definitions.redshift,
+            Definitions.postgres,
+            Definitions.sql_server,
+        }:
+            query += f" FROM {self.database}.{schema_name}.{table_name}"
+        elif self.connection.type == Definitions.bigquery:
+            query += f" FROM `{self.database}`.`{schema_name}`.`{table_name}`"
+
+        return query + ";" if self.connection.type != Definitions.druid else query
+
```
Comment (geometrikz) [inline metrics_layer/cli/seeding.py:395]: This function implements the column cardinality queries.

It takes in a list of column names, constructs the COUNT DISTINCT queries for each column name (names this as {column_name}_cardinality). Then joins all the COUNT_DISTINCT queries into a single query before running it. 

Quotations are added around the column names to handle SQL wildcard characters (e.g. %). BigQuery uses backticks.

**Review by geometrikz — COMMENTED** [review]
Updated to get all cardinalities in single query.

File: metrics_layer/cli/seeding.py [hunk 6]
```diff
@@ -304,12 +328,71 @@ def make_fields(self, column_data: str):
...
+                column_names=searchable_field_candidates, schema_name=schema_name, table_name=table_name
+            )
+            column_cardinalities = self.run_query(query=column_cardinalities_query)
+
+            # Get the column names that have a cardinality of less than 100
+            # Note: running the query doesn't preserve the column name cases
+            searchable_column_names = [
+                col.lower().rsplit("_cardinality", 1)[0]
+                for col in column_cardinalities.columns
+                if column_cardinalities.loc[0, col] < 100
```
Comment (pblankley) [inline metrics_layer/cli/seeding.py:350]: make this settable with the call to seed (e.g. if I want everything searchable with less than 1000 cardinality, let me set that via a argument to the `seed` function)

File: metrics_layer/cli/seeding.py (hunk 5 above)
Comment (pblankley) [reply metrics_layer/cli/seeding.py:395]: This looks good to me, can you confirm you've confirmed the syntax for the db flavors with their docs and tested snowflake end to end?

**Pull request merged**